src/api/wallet/WalletApiImpl.ts

```diff
@@ -0,0 +1,254 @@
+  account: string
+  chainId: number
+  blockHeader: BlockHeader | null
+}
+
+type BlockchainUpdatePromptCallback = (callback: (changedChainData: BlockchainUpdatePrompt) => void) => Command
+
+// provides subscription to blockhain updates for account/network/block
+const subscribeToBlockchainUpdate = ({
+  provider,
+  subscriptions,
+  web3,
+}: {
+  provider: Provider
+  subscriptions?: Subscriptions
+  web3: Web3
+}): BlockchainUpdatePromptCallback => {
+  const subs = subscriptions || createSubscriptions(provider)
+
+  let networkUpdate: (callback: (chainId: number) => void) => Command
+
+  if (isMetamaskSubscriptions(subs)) networkUpdate = (cb): Command => subs.onNetworkChanged(networkId => cb(+networkId))
+  if (isWalletConnectSubscriptions(subs)) networkUpdate = subs.onChainChanged
+
+  const accountsUpdate = subs.onAccountsChanged
+
+  const blockUpdate = (cb: (blockHeader: BlockHeader) => void): Command => {
+    const blockSub = web3.eth.subscribe('newBlockHeaders').on('data', cb)
+    return (): void => {
+      blockSub.unsubscribe()
+    }
+  }
+
+  const {
+    accounts: [account],
+    chainId,
+  } = getProviderState(provider)
+
+  let blockchainPrompt: BlockchainUpdatePrompt = {
+    account,
+    chainId: +chainId,
+    blockHeader: null,
+  }
+
+  const subscriptionHOC: BlockchainUpdatePromptCallback = callback => {
+    const unsubNetwork = networkUpdate(chainId => {
+      blockchainPrompt = { ...blockchainPrompt, chainId }
+      log('chainId changed:', chainId)
+      callback(blockchainPrompt)
+    })
+    const unsubAccounts = accountsUpdate(([account]) => {
+      blockchainPrompt = { ...blockchainPrompt, account }
+      log('accounts changed:', account)
+      callback(blockchainPrompt)
+    })
+
+    const unsubBlock = blockUpdate(blockHeader => {
+      blockchainPrompt = { ...blockchainPrompt, blockHeader }
+      log('block changed:', blockHeader.number)
+      callback(blockchainPrompt)
+    })
+
+    return (): void => {
+      unsubNetwork()
+      unsubAccounts()
+      unsubBlock()
+    }
+  }
+
+  return subscriptionHOC
+}
+
+// const AUTOCONNECT = process.env.AUTOCONNECT === 'true'
+
+const wcOptions: WalletConnectInits = {
+  package: WalletConnectProvider,
+  options: {
+    // TODO get infuraId from .env
+    infuraId: '8b4d9b4306294d2e92e0775ff1075066',
+  },
+}
+
+/**
+ * Basic implementation of Wallet API
+ */
+export class WalletApiImpl implements WalletApi {
+  private _listeners: ((walletInfo: WalletInfo) => void)[]
+  private _provider: Provider | null
+  private _web3: Web3 | null
+
+  private _unsubscribe: Command = () => {}
+
+  public constructor() {
+    this._listeners = []
+  }
+
+  public isConnected(): boolean {
+    return this._connected
+  }
+
+  public async connect(): Promise<boolean> {
+    const provider = await getProvider(wcOptions)
+
+    if (!provider) return false
+
+    if (isMetamaskProvider(provider)) provider.autoRefreshOnNetworkChange = false
```

Comment:
I think we should use this rule: https://eslint.org/docs/rules/curly

It prevents errors, and it's more readable since every line has it's own responsibility:

```js
if (isMetamaskProvider(provider)) { // the check is more clear
   provider.autoRefreshOnNetworkChange = false // the conditonal action is more clear too
}
```

If it's for making code have less lines, I find ok to add the extra lines of the curly braces in these cases. Chaining instructions in the same line is not how we reduce the number of lines more efficiently

Reply:
Sure, add it in a new PR, run linter fix, it'll add itself

Reply:
I created the issue here: https://github.com/gnosis/dex-react/issues/159